File: binders/nodejs/lib/vocabulary.js
// String → uint32 token vocabulary.
//
// Trie keys want short, fixed-width segments, and n-gram contexts want them
// fixed-width by construction. A 40-character sha1 (or any other free-form
// identifier) is neither, so clients end up interning identifiers into small
// integers. This is that allocator, with both directions of the mapping
// persisted.
//
// **Concurrency.** Cheetah has no compare-and-swap, so the counter is guarded
// by a single-flight promise chain in Node. That makes allocation safe within
// one process. Two processes — including two Node worker threads, whose
// CommonJS module state is isolated — allocating into the same database
// concurrently would race and can hand the same id to two names. Either confine
// allocation to one process, or move it server-side.

const { getValue, putValue } = require('./kv');
const { CheetahError } = require('./client');
const { hex } = require('./keys');

/** 0 is reserved as "no token", so ids start at 1. */
const FIRST_TOKEN = 1;
const MAX_TOKEN = 0xffffffff;

const DEFAULTS = Object.freeze({
    counterKey: 'cfg:next_token',
    tokenKey: (name) => `t:${name}`,
    reverseTokenKey: (token) => `r:${hex(token, 8, 'token')}`,
    cacheLimit: 200000,
});

class TokenVocabulary {
    /**
     * @param {object} conn a CheetahClient or CheetahPool
     * @param {object} [options]
     * @param {string} [options.counterKey] where the next free id is persisted
     * @param {function} [options.tokenKey] name → key holding its token
     * @param {function} [options.reverseTokenKey] token → key holding its name
     * @param {number} [options.cacheLimit] in-memory name→token entries
     */
    constructor(conn, options = {}) {
        const config = { ...DEFAULTS, ...options };
        this.conn = conn;
        this.counterKey = config.counterKey;
        this.tokenKey = config.tokenKey;
        this.reverseTokenKey = config.reverseTokenKey;
        this.cacheLimit = config.cacheLimit;
        this.forward = new Map();
        this.reverse = new Map();
        // Serializes every read-modify-write of the counter.
        this.allocationChain = Promise.resolve();
        this.pendingLookups = new Map();
    }

    #remember(name, token) {
        if (this.forward.size >= this.cacheLimit) {
            // Cheapest bounded policy that cannot thrash: drop the oldest insert.
            const oldest = this.forward.keys().next().value;
            const staleToken = this.forward.get(oldest);
            this.forward.delete(oldest);
            this.reverse.delete(staleToken);
        }
        this.forward.set(name, token);
        this.reverse.set(token, name);
        return token;
    }

    async #readCounter() {
        const raw = await getValue(this.conn, this.counterKey);
        if (raw === null) return FIRST_TOKEN;
        const value = Number.parseInt(raw, 10);
        if (!Number.isFinite(value) || value < FIRST_TOKEN) {
            throw new CheetahError(`cheetah ${this.counterKey} holds a non-token value: ${raw}`);
        }
        return value;
    }

    /**
     * Reserve `count` consecutive ids. Serialized against every other caller on
     * this instance; the counter is advanced *before* any mapping is written,
     * so a crash mid-allocation loses ids rather than reusing them.
     */
    async allocate(count = 1) {
        const amount = Math.max(1, Math.trunc(count));
        const run = this.allocationChain.then(async () => {
            const next = await this.#readCounter();
            if (next + amount - 1 > MAX_TOKEN) {
                throw new CheetahError('cheetah token vocabulary exhausted (uint32)');
            }
            await putValue(this.conn, this.counterKey, String(next + amount), { upsert: true });
            return next;
        });
        // Keep the chain alive even when this allocation rejects.
        this.allocationChain = run.then(() => undefined, () => undefined);
        return run;
    }

    /** The token for a name, allocating and persisting on first sight. */
    async tokenFor(name) {
        const cached = this.forward.get(name);
        if (cached !== undefined) return cached;

        const inFlight = this.pendingLookups.get(name);
        if (inFlight) return inFlight;

        const lookup = (async () => {
            const stored = await getValue(this.conn, this.tokenKey(name));
            if (stored !== null) {
                const token = Number.parseInt(stored, 10);
                if (!Number.isFinite(token)) {
                    throw new CheetahError(`cheetah token for ${name} is not numeric: ${stored}`);
                }
                return this.#remember(name, token);
            }
            const token = await this.allocate(1);
            // Reverse first: a reverse entry without a forward one is a readable
            // orphan, while a forward without a reverse breaks every
            // key→name explanation.
            await putValue(this.conn, this.reverseTokenKey(token), name, { upsert: true });
            await putValue(this.conn, this.tokenKey(name), String(token), { upsert: true });
            return this.#remember(name, token);
        })().finally(() => {
            this.pendingLookups.delete(name);
        });

        this.pendingLookups.set(name, lookup);
        return lookup;
    }

    /** The name behind a token, or null when unknown. */
    async nameFor(token) {
        const cached = this.reverse.get(token);
        if (cached !== undefined) return cached;
        const stored = await getValue(this.conn, this.reverseTokenKey(token));
        if (stored === null) return null;
        this.#remember(stored, token);
        return stored;
    }

    /** Resolve many names, preserving input order. */
    async tokensFor(names) {
        const requested = Array.isArray(names) ? names : [];
        const unique = [...new Set(requested)];
        const fresh = unique.filter((name) =>
            !this.forward.has(name) && !this.pendingLookups.has(name)
        );

        if (fresh.length > 0) {
            // Resolve all existing mappings concurrently, reserve one token
            // block for the true misses, then pipeline both mapping directions.
            // Register each name in pendingLookups before the first read can
            // finish so tokenFor() collapses onto this batch.
            const batch = (async () => {
                const storedValues = await Promise.all(
                    fresh.map((name) => getValue(this.conn, this.tokenKey(name)))
                );
                const resolved = new Map();
                const missing = [];
                storedValues.forEach((stored, index) => {
                    const name = fresh[index];
                    if (stored === null) {
                        missing.push(name);
                        return;
                    }
                    const token = Number.parseInt(stored, 10);
                    if (!Number.isFinite(token)) {
                        throw new CheetahError(
                            `cheetah token for ${name} is not numeric: ${stored}`
                        );
                    }
                    resolved.set(name, token);
                });

                if (missing.length > 0) {
                    const first = await this.allocate(missing.length);
                    await Promise.all(missing.flatMap((name, index) => {
                        const token = first + index;
                        resolved.set(name, token);
                        return [
                            putValue(this.conn, this.reverseTokenKey(token), name, { upsert: true }),
                            putValue(this.conn, this.tokenKey(name), String(token), { upsert: true }),
                        ];
                    }));
                }
                return resolved;
            })();

            for (const name of fresh) {
                const lookup = batch
                    .then((resolved) => this.#remember(name, resolved.get(name)))
                    .finally(() => {
                        if (this.pendingLookups.get(name) === lookup) {
                            this.pendingLookups.delete(name);
                        }
                    });
                this.pendingLookups.set(name, lookup);
            }
        }

        return Promise.all(requested.map((name) => this.tokenFor(name)));
    }
}

module.exports = {
    FIRST_TOKEN,
    MAX_TOKEN,
    TokenVocabulary,
    VOCABULARY_DEFAULTS: DEFAULTS,
};

File: binders/python/cheetah_db/vocabulary.py
"""String → uint32 token vocabulary.

Trie keys want short, fixed-width segments, and n-gram contexts want them
fixed-width by construction. A 40-character sha1 (or any other free-form
identifier) is neither, so clients end up interning identifiers into small
integers. This is that allocator, with both directions of the mapping
persisted.

**Concurrency.** Cheetah has no compare-and-swap, so the counter is guarded by a
lock here. That makes allocation safe within one process. Two processes
allocating into the same database concurrently would race and can hand the same
id to two names. Either confine allocation to one process, or move it
server-side.
"""

from __future__ import annotations

import threading
from collections import OrderedDict
from typing import Any, Callable, Sequence

from .client import CheetahError
from .keys import hex_segment
from .kv import get_value, put_value

__all__ = ["FIRST_TOKEN", "MAX_TOKEN", "TokenVocabulary"]

#: 0 is reserved as "no token", so ids start at 1.
FIRST_TOKEN = 1
MAX_TOKEN = 0xFFFFFFFF

DEFAULT_COUNTER_KEY = "cfg:next_token"
DEFAULT_CACHE_LIMIT = 200_000


class TokenVocabulary:
    def __init__(
        self,
        conn: Any,
        *,
        counter_key: str = DEFAULT_COUNTER_KEY,
        token_key: Callable[[str], str] | None = None,
        reverse_token_key: Callable[[int], str] | None = None,
        cache_limit: int = DEFAULT_CACHE_LIMIT,
    ) -> None:
        self.conn = conn
        self.counter_key = counter_key
        self.token_key = token_key or (lambda name: f"t:{name}")
        self.reverse_token_key = reverse_token_key or (
            lambda token: f"r:{hex_segment(token, 8, 'token')}"
        )
        self.cache_limit = max(1, int(cache_limit))
        self._forward: "OrderedDict[str, int]" = OrderedDict()
        self._reverse: dict[int, str] = {}
        self._lock = threading.Lock()

    # -- cache ---------------------------------------------------------- #
    def _remember(self, name: str, token: int) -> int:
        if len(self._forward) >= self.cache_limit:
            # Cheapest bounded policy that cannot thrash: drop the oldest insert.
            oldest, stale_token = self._forward.popitem(last=False)
            self._reverse.pop(stale_token, None)
        self._forward[name] = token
        self._reverse[token] = name
        return token

    # -- allocation ----------------------------------------------------- #
    def _read_counter(self) -> int:
        raw = get_value(self.conn, self.counter_key)
        if raw is None:
            return FIRST_TOKEN
        try:
            value = int(raw)
        except ValueError as exc:
            raise CheetahError(
                f"cheetah {self.counter_key} holds a non-token value: {raw!r}"
            ) from exc
        if value < FIRST_TOKEN:
            raise CheetahError(f"cheetah {self.counter_key} holds a non-token value: {raw!r}")
        return value

    def allocate(self, count: int = 1) -> int:
        """Reserve ``count`` consecutive ids and return the first.

        The counter is advanced *before* any mapping is written, so a crash
        mid-allocation loses ids rather than reusing them.
        """
        amount = max(1, int(count))
        with self._lock:
            first = self._read_counter()
            if first + amount - 1 > MAX_TOKEN:
                raise CheetahError("cheetah token vocabulary exhausted (uint32)")
            put_value(self.conn, self.counter_key, str(first + amount), upsert=True)
            return first

    # -- mapping -------------------------------------------------------- #
    def token_for(self, name: str) -> int:
        """The token for a name, allocating and persisting on first sight."""
        cached = self._forward.get(name)
        if cached is not None:
            self._forward.move_to_end(name)
            return cached
        stored = get_value(self.conn, self.token_key(name))
        if stored is not None:
            try:
                token = int(stored)
            except ValueError as exc:
                raise CheetahError(f"cheetah token for {name!r} is not numeric: {stored!r}") from exc
            return self._remember(name, token)
        token = self.allocate(1)
        # Reverse first: a reverse entry without a forward one is a readable
        # orphan, while a forward without a reverse breaks every key→name
        # explanation.
        put_value(self.conn, self.reverse_token_key(token), name, upsert=True)
        put_value(self.conn, self.token_key(name), str(token), upsert=True)
        return self._remember(name, token)

    def name_for(self, token: int) -> str | None:
        """The name behind a token, or ``None`` when unknown."""
        cached = self._reverse.get(token)
        if cached is not None:
            return cached
        stored = get_value(self.conn, self.reverse_token_key(token))
        if stored is None:
            return None
        self._remember(stored, token)
        return stored

    def tokens_for(self, names: Sequence[str]) -> list[int]:
        """Resolve many names, preserving input order.

        Misses are allocated as one consecutive block rather than one id per
        name: the counter read-modify-write is the expensive part, and doing it
        once per batch keeps a cold ingest from serializing on it.
        """
        requested = list(names or ())
        unique = list(dict.fromkeys(requested))
        missing: list[str] = []
        resolved: dict[str, int] = {}
        for name in unique:
            cached = self._forward.get(name)
            if cached is not None:
                resolved[name] = cached
                continue
            stored = get_value(self.conn, self.token_key(name))
            if stored is None:
                missing.append(name)
                continue
            try:
                resolved[name] = self._remember(name, int(stored))
            except ValueError as exc:
                raise CheetahError(f"cheetah token for {name!r} is not numeric: {stored!r}") from exc
        if missing:
            first = self.allocate(len(missing))
            for offset, name in enumerate(missing):
                token = first + offset
                put_value(self.conn, self.reverse_token_key(token), name, upsert=True)
                put_value(self.conn, self.token_key(name), str(token), upsert=True)
                resolved[name] = self._remember(name, token)
        return [resolved[name] for name in requested]
